### ISSUE TITLE
Enable icecast-kh encryption

### ISSUE THREAD

**Issue by amavarick** [AzuraCast/AzuraCast#358]
@SlvrEagle23 

Feels like I've been submitting alot of tickets lately and have been really self conscious about it.  Trying to be respectful of your workload, sending as many coffees as I can and am so excited about being able to contribute.  I really do try to do my work before just submitting a feature request or bug to properly show the issue/recommendation and think through the process of implementation.  If I knew more, I'd create the pull requests too.  I'm learning though and trying to be helpful.

I ask you to re-evaluate enabling icecast-kh encryption with Azuracast.  icecast-kh has made some drastic improvements to their encryption capability in the last several months.  All dependencies are already installed with AzuraCast docker and it only needs some minor config file updates.  I tested on docker with my implementation and it works great.  Your actual implementation would be smoother as Docker is a weakness of mine at this time.  Either way, its really easy!!!

**Improvements to icecast-kh**
https://karlheyes.github.io/
8/5/2017
--autodetect SSL connections on incoming sockets. No need for <ssl> in listen-socket now but is still there for compatability.  _(THIS IS BIG!!!  We no longer need to use different ports for encrypted and unencrypted dramatically reducing complexity)_
--add ssl-private-key in paths to allow for combined PEM or for separate SSL key/certificate files.  _(THIS IS BIG TOO!  No longer need a separate process after updating let's encrypt to combine your fullchain and private cert into one pem file for icecast to read it!)_
--select https/http URL in autogenerated m3u based on incoming request.

**How to implement on docker**:
1.  Map Let's Encrypt certificate location for icecast-kh (docker-compose.yml?)
2.  Reference your live (letsencrypt or nginx) fullchain and private pem files in your icecast.xml files using mapped location.  Recommend making this a default for all radios that use icecast-kh with let's Encrypt?
3.  Restart appropriate icecast-kh service to activate icecast.xml config.
4.  Update your Let's Encrypt update process to restart icecast-kh services each time you update certificates.  (Based on a trouble ticket, it appears icecast-kh doesn't need service restart to refresh certificates) https://github.com/karlheyes/icecast-kh/issues/162  (May need some testing though as I don't know if responder to that ticket is the author)

**How I implemented on docker because of my lack of docker config knowledge to implement steps above**:
1.  I copied my live fullchain.pem and private.pem files from /var/lib/docker/volumes/azuracast_nginx_letsencrypt_certs/_data/live/myradiodomain.tld/
to
/var/lib/docker/volumes/azuracast_station_data/_data/myradio/config/

2.  I chown user:group to match the other config files in the config directory

3.  I updated the icecast.xml file:
/var/lib/docker/volumes/azuracast_station_data/_data/myradio/config/icecast.xml
Adding the following in the paths section:
```
<ssl-private-key>/var/azuracast/stations/myradio/config/privkey.pem</ssl-private-key>
<ssl-certificate>/var/azuracast/stations/myradio/config/fullchain.pem</ssl-certificate>
```
You can also define the ciphers, otherwise don't include to take application defaults:
`	<ssl-allowed-ciphers>ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:RSA+AESGCM:RSA+AES:!aNULL:!MD5:!DSS</ssl-allowed-ciphers>`

4.  I restarted icecast-kh service

Since icecast-kh allows both http and https off of the same port, no need to update liquidsoap... it will still connect to icecast as it was originally defined.  However, a user can now connect to the stream (same port) either encrypted or unencrypted.  What a Christmas present!!!

For anyone using my steps, a docker update will erase your changes to icecast.xml but the certificates if copied to the config folder will remain.  Also if you update certificates, it will not automatically update the certificates copied to your config folder.

Here is log file showing successful implementation.  I was also able to browse to the icecast page using both http and https as well as connect to the m3u radio location using both http and https. /var/lib/docker/volumes/azuracast_station_data/_data/myradio/config/icecast_error.log

```
[2017-12-25  18:24:36] INFO connection/get_ssl_certificate SSL certificate found at /var/azuracast/stations/myradio/config/fullchain.pem
[2017-12-25  18:24:36] INFO connection/get_ssl_certificate SSL private key found at /var/azuracast/stations/myradio/config/privkey.pem
[2017-12-25  18:24:36] INFO connection/get_ssl_certificate SSL using ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-RSA-DES-CBC3-SHA:ECDHE-ECDSA-DES-CBC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:AES:DES-CBC3-SHA:HIGH:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA
[2017-12-25  18:24:36] INFO connection/connection_setup_sockets 1 listener socket(s) for port 8030
[2017-12-25  18:24:36] INFO connection/connection_setup_sockets 1 listening sockets setup complete
```

**Comment by SlvrEagle23** [comment]
@amavarick I've been exploring this issue in further detail, and I have a few observations:

- Because the nginx container runs as the root user, all of its SSL certificates are, by default, set to being owned by root. Because of the way docker volume mounting works, it's not possible to set per-container permissions that would change these certs to be owned by the `azuracast` user, which is what runs icecast in the stations container. There's a long-outstanding request in Docker's repository to add this, but for now, there's no easy automated way to make it happen.

- Because Shoutcast has no support whatsoever for HTTPS directly, the current solution of automatically proxying all secure connections through nginx is still the most reliable option for now, as it offers uniform support for all radio platforms.

**Comment by amavarick** [comment]
Brainstorming:  Hope this isn't too wordy, just trying to pose options.

So the certificate exists in one container and only root can cross containers?

-It appears that icecast-kh reads the certs as root as the listener socket is setup.
https://github.com/karlheyes/icecast-kh/issues/158
States that the "The SSL file reading is done at the same time as the listener socket setup so should have root reading capability in such configurations."

-Can the certs be symlinked or copied to /var/azuracast/stations/myradio/config/ as I have done above?  Then append the let's encrypt update script to copy the cert over if symlink can't be used?

Maybe a toggle switch if Icecast is selected instead of Shoutcast.  That toggle switch could be, "Enable SSL with Icecast" and upon selection.  A back-end script copies the privkey and fullchain over to the /var/azuracast/stations/myradio/config/ folder.  Then it appends the icecast.xml with the the certificate and private key entry.  Then update your let's encrypt updating script to copy the certs over to each radio upon renewal.

Its my understanding that this can work with proxy enabled as to use the icecast URL we'd be using the icecast port and not the AzuraCast nginx proxy url?  In otherwords, with my configuration posted previously, I can connect to the icecast port and play directly OR use the proxy.  Specifically because Icecast doesn't require you to change your icecast port to use http or https.  It can be done from same port so Azuracast can connect as it needs and do what it wants while I can also connect to icecast port directly without using the enabled proxy.  My implementation works inline to yours whether proxy is enabled or not.  The only issue for me is surviving AzuraCast updates because it re-writes the icecast.xml file but /config folder and contents remain.  The other issue is updating the certificates after they're updated on the nginx side.

**Comment by SlvrEagle23** [comment]
@amavarick From my testing, the fact that the SSL certs were owned by root caused a problem, which resulted in an error log message that said the certificate could not be loaded, and thus SSL was unavailable on all incoming connections.

Right now, as long as Icecast-kh fixes the issue with the "X-Forwarded-For" header not being respected on Docker, there's absolutely no disadvantage to relying on the nginx proxying for secure connections, which offers a high performance throughput and adds no additional latency as the proxying is entirely via local network sockets.

**Comment by amavarick** [comment]
Just validated, the certs can remain as root even when the service start as another account:

![image](https://user-images.githubusercontent.com/20918832/34447608-76138e72-ecb3-11e7-839d-6afcb46e8701.png)

![image](https://user-images.githubusercontent.com/20918832/34447631-9fafe58c-ecb3-11e7-860a-b513b4ae2adb.png)

**Comment by SlvrEagle23** [comment]
@amavarick Turns out there was a minor issue in how the self-signed cert was being generated (it was being set as valid for 3,650 days, not 365, and this made icecast's validation of it flop). Once that was fixed, the native SSL worked on Icecast just fine.

I've updated the Docker configuration to mount the SSL certificates volume to the stations container as read-only, and updated the configuration generator to refer to the proper paths to make everything work for both the self-signed and letsencrypt cert varieties.

The station URL generation has also been updated to use the direct HTTPS URL if "use radio proxy" is set to "no" for an Icecast station, and either you're viewing the page via HTTPS or "Always Use HTTPS" is set to "yes".

Try it out.

**Comment by amavarick** [comment]
Oh that's great to hear!  During testing, I got the following...

My settings:
![image](https://user-images.githubusercontent.com/20918832/34473368-c1454b22-ef3f-11e7-8c42-8f0bec1d4b96.png)

icecast.xml settings:
![image](https://user-images.githubusercontent.com/20918832/34473347-84ea1432-ef3f-11e7-8047-65a075a63b9d.png)

icecast_error.log
![image](https://user-images.githubusercontent.com/20918832/34473327-4f13a7ec-ef3f-11e7-934c-bca19df1e091.png)

I also tried the following with same results:
![image](https://user-images.githubusercontent.com/20918832/34473433-759e0078-ef40-11e7-81eb-e3598730dcbf.png)

I restarted each time to be sure there was no issue with a service needing restarted.

For this to work, should they reference fullchain.pem and privkey.pem in icecast.xml?

**Comment by SlvrEagle23** [comment]
@amavarick There are symlinks in place to ensure the same configuration works with both self-signed certs and the letsencrypt certs, so `/etc/nginx/ssl/ssl.key` and `/etc/nginx/ssl/ssl.crt` will work as valid paths.

I'm seeing the same issue myself, that for whatever reason this setup isn't working with LetsEncrypt certs, though it's now working with self-signed certs.

Icecast is unfortunately very lacking in verbosity with this kind of issue, so it'll need further exploration.

**Comment by SlvrEagle23** [comment]
@amavarick Closing this issue as
A) I can't seem to figure out what on earth is going wrong with our current setup, and
B) Depending on Xiph's response to both your ticket and mine in their main-branch Icecast Github, we may be switching over to the main branch again. This will change the method of implementation for this, and we will have to revisit it once that happens. I don't want to invest too much time into a KH-specific solution in the meantime, as nginx proxying does accomplish the purpose.

**Comment by SlvrEagle23** [comment]
@amavarick Just wanted to let you know that as part of the changes to our Docker container infrastructure, including the rollout of our "Unified" docker container that merges the functionality of 5 previous containers into a single one, I also made it a point to ensure that certbot now properly generates certificates under the `azuracast` user instead of root. This fixes the secondary issue of making the SSL certs work without issue on Icecast.

If you want to take advantage of these new improvements, I would recommend updating both your `docker.sh` script (`docker.sh update-self`) and your `docker-compose.yml` file.
